### PR TITLE
zip_longest lists and tuples, batches can be sized differently

### LINKED ISSUE #25939: nested_concat assumes lists are equal in size
### System Info

I have stumbled upon a quirk while trying to figure out how to calculate custom metrics.

using a detr model for object detection and a provided trainer with a dataset with that has a smaller last batch I am missing labels in the custom metric input.

the length of `batched_labels` in the metric is per the length of the last batch (smaller) and isn't merged like the other fields via the https://github.com/huggingface/transformers/blob/0afa5071bd84e44301750fdc594e33db102cf374/src/transformers/trainer_pt_utils.py#L105 on line  
https://github.com/huggingface/transformers/blob/0afa5071bd84e44301750fdc594e33db102cf374/src/transformers/trainer.py#L3237
for equal sized batches it works fine. (total eval_dataset size is 16, batched automatically into 4's)
image

image

### Who can help?

@muellerz @pacman100

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

setup similar to: https://huggingface.co/docs/transformers/main/en/tasks/object_detection#object-detection

```
def compute_metrics(eval_pred: EvalPrediction): 
    (loss_dict, logits, pred_boxes, last_hidden_state, encoder_last_hidden_state), batched_labels = eval_pred
    # (loss_dict, logits, pred_boxes, last_hidden_state, encoder_last_hidden_state), labels = eval_pred
    outputs = DetrObjectDetectionOutput(#loss_dict=loss_dict,
                                        logits=torch.from_numpy(logits),
                                        pred_boxes=torch.from_numpy(pred_boxes),
                                        last_hidden_state=None,
                                        decoder_hidden_states=None,
                                        )
    number_of_image_ids_in_each_indecies = [batched_label['image_id'].shape for batched_label in batched_labels]
    print(number_of_image_ids_in_each_indecies)
 ```

trainer,
```
trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=collate_fn,
    train_dataset=ds_train_augmented,
    eval_dataset=ds_val_augmented,
    tokenizer=image_processor,
    compute_metrics=compute_metrics,
)
```

### Expected behavior

to pad it to the last? to append it to a growing list?
anything but this that retains the data.

### PR DESCRIPTION
**Pull request by YoraiLevi**
# What does this PR do?

Fixes #25939
image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [❌] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [❌] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [❌] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [❌] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @muellerzr @pacman100

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3184,7 +3184,8 @@ def evaluation_loop(
 
             if labels is not None:
                 labels = self.accelerator.gather_for_metrics((labels))
-                labels_host = labels if labels_host is None else nested_concat(labels_host, labels, padding_index=-100)
+                labels_host = [labels] if labels_host is None else labels_host+[labels]
```
Comment (muellerzr) [inline src/transformers/trainer.py:3187]: Why are we no longer using nested_concat here?

Comment (YoraiLevi) [reply src/transformers/trainer.py:3187]: Because of https://github.com/huggingface/transformers/issues/25939#issuecomment-1704588237
This is a botch to make my evaluation work and keep a list (of batches) of lists (of examples) separated.
As it is now they're merged by `nested_concat` in a way that is irreversible otherwise.
It is likely that the solution should be revisiting `nested_concat`

Comment (YoraiLevi) [reply src/transformers/trainer.py:3187]: run with a gpu; looks at last cell's output
https://colab.research.google.com/drive/17VzhS0rlswL-zmd1wPZXSRfFUsGDLfWB?usp=sharing

Comment (muellerzr) [reply src/transformers/trainer.py:3187]: If your solution works for your problem, we definitely should consider it. But we need to make sure this doesn't considerably break things existing (see the current failing tests) :)

Comment (YoraiLevi) [reply src/transformers/trainer.py:3187]: > If your solution works for your problem, we definitely should consider it. But we need to make sure this doesn't considerably break things existing (see the current failing tests) :)

I definitely agree! I am not sure what your preferences are at huggingface.
it is quite confusing the entire way it handles input to the metrics, it seems like you're handling the entire model output into the metric with some intention in mind, tuplizing it and gathering batches etc etc. I am not sure what to suggest as an actual alternative at this time because of that.
it's just impossible to calculate any metric with that detr output as it is now...
